h2/src/main/org/h2/table/TableSynonym.java

```diff
@@ -0,0 +1,193 @@
+        getSynonymFor().close(session);
+    }
+
+    @Override
+    public void unlock(Session s) {
+        getSynonymFor().unlock(s);
+    }
+
+    @Override
+    public Index addIndex(Session session, String indexName, int indexId, IndexColumn[] cols, IndexType indexType, boolean create, String indexComment) {
+        throw DbException.getUnsupportedException("SYNONYM");
+    }
+
+    @Override
+    public void removeRow(Session session, Row row) {
+        getSynonymFor().removeRow(session, row);
+    }
+
+    @Override
+    public void truncate(Session session) {
+        getSynonymFor().truncate(session);
+    }
+
+    @Override
+    public void addRow(Session session, Row row) {
+        getSynonymFor().addRow(session, row);
+    }
+
+    @Override
+    public void checkSupportAlter() {
+        throw DbException.getUnsupportedException("SYNONYM");
+    }
+
+    @Override
+    public TableType getTableType() {
+        return TableType.SYNONYM;
+    }
+
+    @Override
+    public Index getScanIndex(Session session) {
+        return getSynonymFor().getScanIndex(session);
+    }
+
+    @Override
+    public Index getUniqueIndex() {
+        return getSynonymFor().getUniqueIndex();
+    }
+
+    @Override
+    public ArrayList<Index> getIndexes() {
+        return getSynonymFor().getIndexes();
+    }
+
+    @Override
+    public boolean isLockedExclusively() {
+        return getSynonymFor().isLockedExclusively();
+    }
+
+    @Override
+    public long getMaxDataModificationId() {
+        return getSynonymFor().getMaxDataModificationId();
+    }
+
+    @Override
+    public boolean isDeterministic() {
+        return getSynonymFor().isDeterministic();
+    }
+
+    @Override
+    public boolean canGetRowCount() {
+        return getSynonymFor().canGetRowCount();
+    }
+
+    @Override
+    public boolean canDrop() {
+        return false;
+    }
+
+    @Override
+    public long getRowCount(Session session) {
+        return getSynonymFor().getRowCount(session);
+    }
+
+    @Override
+    public long getRowCountApproximation() {
+        return getSynonymFor().getRowCountApproximation();
+    }
+
+    @Override
+    public long getDiskSpaceUsed() {
+        return getSynonymFor().getDiskSpaceUsed();
+    }
+
+    @Override
+    public String getCreateSQL() {
+        return "CREATE SYNONYM " + getName() + " FOR " + data.synonymForSchema.getName() + "." + data.synonymFor;
+    }
+
+    @Override
+    public String getDropSQL() {
+        return "DROP SYNONYM " + getName();
+    }
+
+    @Override
+    public void checkRename() {
+        throw DbException.getUnsupportedException("SYNONYM");
+    }
+
+    public boolean canTruncate() {
+        return getSynonymFor().canTruncate();
+    }
+
+    public String getSynonymForName() {
+        return data.synonymFor;
+    }
+
+    public boolean isInvalid() {
+        try {
+            getSynonymFor();
+            return false;
+        } catch (DbException e) {
```

Comment:
Changed to:  return data.synonymForSchema.findTableOrView(data.session, data.synonymFor) != null;